src/vos/vos_aggregate.c

```diff
@@ -880,6 +903,8 @@ insert_segments(daos_handle_t ih, struct agg_merge_window *mw,
 
 	/* Insert new segments into EV tree */
 	for (i = 0; i < io->ic_seg_cnt; i++) {
+		if (io->ic_segs[i].ls_has_csum_err)
+			continue;
```

Comment:
Skip only the logical segment with csum error? I think it's unnecessary complicate and error prone. For instance, there are two overlapping physical entries, one of them has checksum error, how will you handle this?

Shouldn't we just abort the whole merge-window (or the whole evtree aggregation)?

Reply:
I suggest to skip this whole evtree aggregation for checksum error for this moment.

Reply:
More extreme, and maybe not achievable, to skip aggregation for the entire evtree, since some insertions could already have been completed in prior windows.  The solution implemented was agreed upon in a discussion held with Johann and Liang at SC.  For the overlap, the holdover is set to the media offset size of prior output range, along with associated csum info.  This csum info will be invalid, since the failure on the prior physical entry will abort the recalulation for the carry-over extent.  Hence all output of new segments based the errant physical extent should be aborted.  However, it may be better to abort the whole window.

Reply:
It's not trivial to make it (aborting segment only) correct, because:
1.  Let's assume two overlapping physical entries. [0, 100) at epoch 2, [50, 150) at epoch 1, the [50, 150) has checksum error, so in your current implementation, [0, 100) will be deleted but [50, 150) will be retained. This is not correct;
2. Aborting segment only requires PMDK and VEA to support "partial publish & partial cancel in one transaction", otherwise, you won't be able to properly cancel the resource reserved by the aborted segments;

My suggestion is to at least abort whole merge-window (either commit or abort the whole merge-window flush transaction), even we go for aborting whole merge-window, we still need to think it over to make sure the extent spans window won't cause any problem.

Reply:
To skip physical entry with checksum error, I think the proper way is to exclude them out during prepare & fill phase (for example, split or truncate the segment which contains data from bad physical entry in fill phase), so that in the final phase (insert phase), all the prepared segments should be good.

Reply:
Yes, I agree with Niu's proposal, we should just skip the corrupted extent, the corrupted data can be reconstructed by scrubbing in the future.